### ISSUE TITLE
The ability to customize tabs should be considered

### ISSUE THREAD

**Issue by chengjianhua** [Microsoft/vscode#41909]
Feature Request

The Atom has the ability to customize the tabs, it's so powerful for some requirements. For example, there is an extension [nice-index](https://github.com/joshwcomeau/nice-index) in Atom that could be able to rename tabs of many opened editors have same name `index.js`, rename them to the names of their parent folder.

It's very useful that if we could do so. 

@bpasero I got it that the feature requested by #32836 will not be considered in the first half year of 2018. I think my feature request is very similar with that, so may I ask you when the ability to customize tabs in VSC would be added into the roadmap ?

I was excited when I start to think how can I implement a extension like `nice-index` in VSC, but after I involved in the documents about customizing a VSC extension, I just found that I can not achieve this. I hope that you could consider this feature request, thanks for all of your great works ~ 🙏

**Comment by bpasero** [comment]
@chengjianhua can you give more details about what you would want to change on the tabs from your extension?

**Comment by chengjianhua** [comment]
@bpasero I want to there is an api that I can leverage it to get the file paths of all active editors, and then use the file paths to generate the custom labels for active editors.

According to the `setting.json` in VSC, I guess this is similar with `workbench.editor.labelFormat`, but more customizable.

https://github.com/Microsoft/vscode/blob/204f0482da3e887ad4a852a87aa9ab6079aefda4/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L380

https://github.com/Microsoft/vscode/blob/204f0482da3e887ad4a852a87aa9ab6079aefda4/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L292

**Comment by usernamehw** [comment]
You want to create an API for a single setting? Wouldn't it be easier to just extend the list with another option  for `"workbench.editor.labelFormat"` - something like: 
```javascript
"workbench.editor.labelFormat": "onlyParent"
```

**Comment by chengjianhua** [comment]
@usernamehw the `parent` in `workbench.editor.labelFormat` does not mean the name of parent folder. 

If you think that my this requirement could be covered by adding an option in `workbench.editor.labelFormat`, I'd like to accept that. For further scalability, providing a interface for this would be better. For example, the aforementioned image shows that there is an slash `/` before the name of parent folder that indicates this is a `index.js` file  in `/Parent` folder.

If I want change the  slash `/` to other symbols, the api is needed. The options are preset for normal users, api for users that want more customizations and implement some interesting ideas.

**Comment by DanTup** [comment]
This came up for Dart too: https://github.com/Dart-Code/Dart-Code/issues/577

When a user F12's into a file from an external library, the filename is often fairly vague. It'd be nice if we could insert the packages name into the tab title text.

**Comment by ncallaway** [comment]
I just ran into the same exact issue with as the parent poster. I was having the same usability problem with duplicated index.js files, and went looking for a solution. Finding no VSC extensions that would help in the marketplace I dug into the documentation along the exact same lines as chengjianhua. I too was disappointed to find that it was not possible to update the content of the editor tabs in any way from the extension API.

I also have the same issue to a lesser extent in rust when dealing with mod.rs files.

The current functionality provided by VSC of showing the folder in a de-emphasized label is helpful, but not as helpful as it could be, as I actually care _more_ about the folder name than the file name in these cases.

What I'd like to be able to modify through an extension is:
* The text-content of the _title_ label of the tab (`index.js` in the screenshot; this is the highest priority item)
* The text-content of the _hint_ label of the tab (`.../SharedHeader` in the screenshot; this would be a secondary priority, but I can't imagine it's that much more work than providing API access to the title label)
* The icon displayed in the tab (`JS` in the screenshot; this is personally a low-priority for me, and I'm including it more for completeness)

![editor_tabs](https://user-images.githubusercontent.com/427497/37377528-b3d68c46-26e7-11e8-91e6-35cb40332d18.png)

**Comment by karlhorky** [comment]
Found this issue after extensive searching, and it looks like a good option for allowing being able to control the tab titles and labels.

This is a big problem in Node projects, where `index.js` files are all over the place. My editor often looks like this:

<img width="873" alt="screen shot 2018-03-23 at 13 18 37" src="https://user-images.githubusercontent.com/1935696/37828882-40741ee0-2e9d-11e8-85ae-b219639941d1.png">

Alternately, if an open, extensible API is too much to consider now, I would suggest re-considering #30316, which also provides a solution to the `index.js` problem.

**Comment by karlhorky** [comment]
Found this issue after extensive searching, and it looks like a good option for allowing being able to control the tab titles and labels.

This is a big problem in Node projects, where `index.js` files are all over the place. My editor often looks like this:

<img width="873" alt="screen shot 2018-03-23 at 13 18 37" src="https://user-images.githubusercontent.com/1935696/37828882-40741ee0-2e9d-11e8-85ae-b219639941d1.png">

Alternately, if an open, extensible API is too much to consider now, I would suggest re-considering #30316 (related to #27328), which also provides a solution to the `index.js` problem.

cc @bpasero @sohlex @rosyatrandom

**Comment by karlhorky** [comment]
Opening this up would also allow for things like @comerc's https://github.com/Microsoft/vscode/issues/12965#issuecomment-326016306 or @aazarkhish's issue https://github.com/Microsoft/vscode/issues/21662

**Comment by karlhorky** [comment]
Also support from @8balloon, @silesky and 5 others for the `index.js` use case here:

https://github.com/Microsoft/vscode/issues/12965#issuecomment-267773692

**Comment by DanTup** [comment]
In Dart it's common go import a file like `package:mypackage/somefile.dart`. If you F12 into that file, the file system location is buried away in some pub cache folder somewhere and the user only sees `somefile.dart`. I'd love ot be able to show `package:mypackage` in the place of the long filesystem path (whether it's on the tab, or a tooltip).

**Comment by tomByrer** [comment]
I hope this feature gets added.  It is not uncommon for a ReactJS project to have 20+ `index.*` files.   I was wondering why `nice-index` had not been ported yet.

**Comment by tomByrer** [comment]
@bpasero  The nice-index author switched to VSCode also, & is awaiting the API ;)

**Comment by datoml** [comment]
That's great news. I hope they implement it soon.

**Comment by AJTJ** [comment]
I am definitely waiting on this to be real. It would be tremendously useful.

**Comment by chengjianhua** [comment]
@bpasero Hi ~ Is there any new chances to plan this?

**Comment by adrian-amigues** [comment]
Just started working on a new codebase with only index.js files... This would be tremendously nice to have!

**Comment by AJTJ** [comment]
don't wanna be annoying, but anyone working with the Component Folder Pattern in React (which is a LOT of people) will really appreciate this.

**Comment by benhancockco** [comment]
Just stumbled across this issue. Exactly what I've been looking for. I'm dealing with heaps of React components all named `index.js`, and this would be a life saver.

**Comment by Lsystems** [comment]
Same issue for me, the best could be able to change the color and the label of a tab on right click.

**Comment by ackvf** [comment]
If a proper API is way too difficult for now, I propose that there is an built-in mechanism that triggers specifically for `index.` files. https://github.com/Microsoft/vscode/issues/67974

**Comment by karlhorky** [comment]
This issue is exacerbated with folders with prefixes in them. Here I cannot tell without hovering which `index.mdx` is in which folder:

<img width="792" alt="Screen Shot 2019-03-21 at 13 43 38" src="https://user-images.githubusercontent.com/1935696/54752935-0202b480-4be0-11e9-8854-947d45c5bd64.png">

@bpasero (or happy to redirect to someone else responsible for the tab bar), is there any initiative to do anything about this?

Even a simple default-off option (show directory name as primary label, when there are duplicates) would be enough for a start...

**Comment by johnathanz** [comment]
Is this a duplicate of https://github.com/microsoft/vscode/issues/37414?

**Comment by ncallaway** [comment]
@johnathanz 

I don't think this is a duplicate. #37414 is asking for a very specific feature be built into VSCode, while this issue is asking for an extension API.

If this issue were implemented, I imagine an extension could provide the functionality requested in #37414 without it needing to be built into VSCode.

As such, I think they are wholly distinct issues, and I would strongly urge _this_ issue to be implemented over #37414, as it would then allow the community to address #37414.

**Comment by johnathanz** [comment]
That makes sense. Thanks for the clarification @ncallaway

**Comment by mohyeid** [comment]
This can open up a lot of possibilities. Such extension and lots of other ideas would be super helpful.

**Comment by BrightSoul** [comment]
Please add this feature, I have too many index files!
![index](https://user-images.githubusercontent.com/1888641/72145892-b6520f00-339b-11ea-8b35-c62ec7b6e71a.png)

Some regexp setting would be much appreciated to give perfect control to the tab name. Something like this would be perfect.
```
{
  "workbench.editor.labelName": {
      "match": "components/(.+?)/index\\.(.+)"
      "replaceWith": "$1.$2"
   }
}
```
Thanks

**Comment by xpl** [comment]
Any update? I desperately want that feature!

**Comment by RobKohr** [comment]
@BrightSoul Yeah, I was having this problem. If we could use regex to set labelnames, that would work great too.

**Comment by RobKohr** [comment]
> I hope this feature gets added. It is not uncommon for a ReactJS project to have 20+ `index.*` files. I was wondering why `nice-index` had not been ported yet.

Yeah, for React projects instead of calling a file ComponentName/index.js I call it ComponentName/ComponentName.js to get around this very problem. It does make import paths longer, but it is a reasonable trade off to get tab labels to be meaningful.

**Comment by karlhorky** [comment]
@RobKohr that is fine for a project that you own and control, but there are many, many codebases with this convention that you will not be able to change.

So this is not a complete workaround for many use cases.

**Comment by TheLogan** [comment]
Any news on this API update?

**Comment by b-d-m-p** [comment]
Please add this. 🙇

**Comment by martis-git** [comment]
Still actual!

**Comment by joaomendoncaa** [comment]
Well, after discovering I could "alt + [number]" to switch between tabs I didn't stop until I could make the tabs display their index. 
I know this is a very niche request so after not finding any solution for this I considered making my own extension for this. Would be a perfect first extension for VSCode.

And now I stumbled upon this...

This is my vote to make the Tabs API solve all the above issues.

**Comment by leejoramo-d51** [comment]
Hard to believe that this issue has been here for so long considering how common of an issue it is. Javascript and Rust have been mentioned above, but here is a regex off the top of my head for the types of files I regularly deal with:

```regex
\.?(ENV|index|setup|config|app|README|TODO)\.?(js|py|php|html?lmd|txt)
```

And those are just the ones that came to mind.

I remember dealing hundreds of `index.html` files with this in the late 1990's with large static site projects. Today a large vscode *workspace* for a node.js project can lead to many duplicate file names too.

**Comment by adamerose** [comment]
Would love to see this implemented.

FYI for anyone else working with lots of `index.js` files, you can get the directory name to show up without the `.../` prefix using the `workbench.editor.labelFormat` setting:

![image](https://user-images.githubusercontent.com/10884874/129462037-c1ee26ad-e6d4-4946-988e-6d3a7fb274e5.png)

**Comment by abhijit-hota** [comment]
> Well, after discovering I could "alt + [number]" to switch between tabs I didn't stop until I could make the tabs display their index.
> I know this is a very niche request so after not finding any solution for this I considered making my own extension for this. Would be a perfect first extension for VSCode.
> 
> And now I stumbled upon this...
> 
> This is my vote to make the Tabs API solve all the above issues.

Crazy how we have had exactly the same experience. I wanted to make this too. I use a similar extension that displays numbered tabs in the browser and I thought someone would have definitely made this for VSCode. When I didn't find any such extension I thought to make my own because, it "would be a perfect first extension" and that's how I ended up here.

Not adding a lot to the discussion but seconding a very valid use-case.

**Comment by mathiasvalentin-thg** [comment]
How is this not added yet?! Working with index files is a mess even with the short label format... 😥

**Comment by azinit** [comment]
Still actual!!! 🤷‍♂️

**Comment by maxfahl** [comment]
Looking forward to be able to see what files I have opened (y)

**Comment by ivangretsky** [comment]
Need this badly too. In php context.

**Comment by tkdennis** [comment]
Still hoping for editor tabs API!

**Comment by DavidLGoldberg** [comment]
I would love this for my extension, [Jumpy2](https://marketplace.visualstudio.com/items?itemName=DavidLGoldberg.jumpy2).  I have had this functionality working in [Jumpy](https://atom.io/packages/jumpy) for Atom for a long time.  Ideally I'd be looking for a way to use a decorator, and or apply some css access, so I could do some text via "content" attribute.

**Comment by YoussefZidan** [comment]
any updates from 2022 ??

**Comment by andreas-ateles** [comment]
Has there been any official stance or comment on this? This feature would really be useful.

**Comment by Hunkoys** [comment]
I can't wait to have this feature. Together with the ability to sort index files first in a directory. It will be a historical moment.

**Comment by TheLogan** [comment]
It's been 4 years, how is there no news on this? xD

**Comment by kamilmac** [comment]
C'mon MS. You can do that!

**Comment by perpetualgrimace** [comment]
Gimme

**Comment by eIiot** [comment]
Anytime...

**Comment by letsandeepio** [comment]
IMO should be easy to expose the API for changing the tab label, I would want to be able to display the tab number so I can easily switch using keyboard shortcuts.

**Comment by arxpoetica** [comment]
Fair warning, the Svelte team is about to descend on this issue...

![image](https://user-images.githubusercontent.com/111424/181831486-af0e5d3f-6b3b-4569-a6b7-9bd84a8e152c.png)

We have an open discussion (https://github.com/sveltejs/kit/discussions/5748) to rename ALL routed `*.svelte` files to `+page.svelte` and `+layout.svelte` and so forth, so repeated file names are going to be a big thing in Svelte.

**Comment by hyunaki** [comment]
I'm using vscode to work with GameMaker and it become unmanageable very quickly because every tab label is the same after opening some object events (Create_0.gml, Step_0.gml, Step_1.gml, Draw_2.gml...). Hopefully this feature will arrive soon to make my life easier...

**Comment by a15ade** [comment]
I'm actually amazed after scrolling so much, not to find a message announcing implementation of this so common issue

**Comment by atwoodhouse** [comment]
Just finished migrating a project to ultra modern SvelteKit, with the new plus sign routing. I thought I would meet an editor tab label nightmare after the migration, but I didn't. Somehow the file tabs in vscode have the directory name of the `+server.ts` and `+page.svelte` files in a smaller, semi transparent font after the file name. And it only shows for those + route files, without cluttering other more verbosely named files.

![image](https://user-images.githubusercontent.com/11442871/185766683-8cc2dd95-ae8c-4762-bc89-0fc659933a44.png)

Went to revisit this issue, thinking that I might find an explanation to what's going on here, but this issue is still unresolved. After playing around a bit, this seems to be the native behaviour of vscode when having two files open with the exact same file name. Good news for all SvelteKit folks!

However, I still hope that we'll get a proper API for accessing and changing the labels. This only solves it for cases like SvelteKit where the exact same filename is used over and over, and not like @hyunaki's case with GameMaker above. And even for SvelteKit it's a bit annoying that you don't get the directory name appended if you don't happen to have multiple +server or +page files open at once.

**Comment by canvaspixels** [comment]
Please MS, would love this API, it'd be a game changer, this project has so many pointless index.ts files

**Comment by Nesvet** [comment]
Last Christmas you gave me those tabs 🎅

**Comment by tkdennis** [comment]
And the very next day, you took them away 🌨️

**Comment by chutnee** [comment]
This year, *Inhales Copium* :frog:

**Comment by kaermorchen** [comment]
In the meantime 2023 year started

**Comment by John-Di** [comment]
🥲

**Comment by nickyhajal** [comment]
It's incredible how helpful it will be to finally have this - hope it's still coming down the pipe.

**Comment by Anton256** [comment]
Vote up. This feature would be very usefull!

**Comment by mechatroner** [comment]
When this gets implemented, please consider adding a parameter to `openTextDocument(options?: {content: string, language: string}): [Thenable]` function e.g. `label: string`.

**Comment by benmccann** [comment]
@lramos15 if we were to put together a PR for this, would the best way to do that just be to send the PR? Or do you have any up-front thoughts on the API design?

This is a particularly acute problem for SvelteKit where there are a large number of `+page.svelte` files (https://kit.svelte.dev/docs/routing#page-page-svelte). There's currently a post on the front page of HN about SvelteKit which is overall quite positive, but the one recurring complaint people have is regarding finding `+page.svelte` files in VS Code, so we'd love to address it: https://news.ycombinator.com/item?id=36428525

**Comment by benmccann** [comment]
It looks like there are multiple file extensions to one language identifier and one identifier to multiple extensions. And one extension probably can register multiple providers. But there are probably existing features that only execute one provider. Hover allows multiple, but signature help is first non-null.

@lramos15 would it make sense for this feature to utilize the first non-null tab label provided by an extension similar to signature help?

**Comment by leejoramo-d51** [comment]
> It could also possibly be an option of the Svelte extension to use emojis instead for brevity - though I find this to be less clear

I think an option for emojis or your own abbreviations would be a great feature.  As someone who tries to balance long and expressive filenames, with readability in tabs, I too am uncertain about how useful they would be to me, but I would certainly try it.

**Comment by joaomendoncaa** [comment]
I just wanted to add that everyone which left a comment here since 2018. We'll make history when we finally get access to manipulating the tabs!

At this point this is already a race between us getting access to an API vs AI making "auto generative UI" possible where we can all edit the whole interface of vscode through voice on demand LOL 😄

**Comment by Luke-Callaghan23** [comment]
I want to note that not all these features need to be added right away.  Just having the ability to access the `label` field of a workbench tab in extensions would be great as I'm sure the extension community could implement all extra features here, once given the ability to do so.  So many environments/frameworks would benefit greatly from this 🙏

**Comment by benmccann** [comment]
The location that would need to be edited is here:

https://github.com/microsoft/vscode/blob/37aed2e103024d8992fe42cd95aded33db26e502/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1133

(This was also mentioned in one of the first comments above https://github.com/microsoft/vscode/issues/41909#issuecomment-359251523, but I wanted to share an updated permalink since the link above while still fairly relevant is to a much older version of the code)

**Comment by benmccann** [comment]
Today the labels have a bolder name and lighter description. 

![Screenshot from 2023-09-12 14-40-37](https://github.com/microsoft/vscode/assets/322311/d416ce47-afb6-4683-8db4-03a6975bc94b)

The easiest way to update this would be to keep the name and description properties in tact, but change how they're populated. E.g. we could put everything in the name:

![Screenshot from 2023-09-12 14-42-57](https://github.com/microsoft/vscode/assets/322311/4aa3d3fb-a64d-430b-8739-4054c682a65e)

![Screenshot from 2023-09-12 14-59-59](https://github.com/microsoft/vscode/assets/322311/a5c184e8-2371-4f82-aec6-140e060c5b24)

Perhaps nicer looking, but potentially not as easily scannable, would be to swap the contents of the name and description:

![Screenshot from 2023-09-12 14-44-04](https://github.com/microsoft/vscode/assets/322311/88d07845-89fd-432f-adfc-e8b8547c7091)

![Screenshot from 2023-09-12 14-57-11](https://github.com/microsoft/vscode/assets/322311/294bfd5c-677c-4c69-a3bf-3a6a49f4ed73)

I kept the `.../` in for now to demonstrate what it would look like with nested paths, but for files within 1 level of `routes` we could skip that. E.g. just show `faq` rather than `.../faq`.

If the Svelte community had a strong preference for one of these we could potentially just introduce new `editor.labelFormat` option rather than updating the API. If we did want to update the API, I think the easiest way would be to let extensions modify the existing label data structures with name and description rather than doing something more custom like returning arbitrary HTML.

**Comment by leejoramo-d51** [comment]
@benmccann I prefer the later version
*  **../search** +page.svelte

**Comment by benmccann** [comment]
@lramos15 in our investigation we found there's a second place the tab labels are displayed, which is the "Open Editors" view. I think it probably makes sense not to update "Opened Editors" because that view _always_ shows the filename followed by full directory path - the "Editor: Label Format" setting never affects it. Given that the "Opened Editors" view cannot display different label formats today, I feel it might make the display too inconsistent if we had extensions display the labels there in a different manner, but the labels were not affected by existing settings.

![Screenshot 2023-10-04 at 10 52 45](https://github.com/microsoft/vscode/assets/322311/40d79b3f-5dfd-440f-b68e-be25c612ac46)

Our plan right now is to PR something for this that would affect only the editor tab labels and not the "Open Editors" view. Please advise if you agree with this approach or have different thoughts.

**Comment by Alex0007** [comment]
Hi @lramos15 

I'm trying to work on this, but I have some question.

Currently tab labels are calculated from `editor.getName()` method. While it is using `private readonly name` or `private readonly _name` variables internally in different implementations to store internally calculated value (for example value is calculated depending on opened resource via labelService)

Lets say we want ability to set custom labels for tabs. Should we interfere with editor names, or better introduce new property/method with intention to store editor tab label information. 

My preference is to introduce new separate fields and do not interfere with `editor.getName()`. But I'm not sure

**Comment by geekley** [comment]
It's important that we should also be able to dynamically set any icon URI for tabs in custom editors.
A custom editor may want to change an icon arbitrarily or only be able to set an appropriate icon after analyzing the contents of the file. This should work like an override - setting it back to null should restore whichever theme icon would normally be used on the tab.

On that note, there should also be a declarative way to set the icon for custom editors on package.json, like you can set an icon (fallback) in contributed languages.
This isn't done properly even in VSCode built-in views - I can't believe views like the Settings Editor (UI) and Shortcut Editor still use a generic "file" icon instead of something like a cogwheel/wrench/etc.

**Comment by wesbos** [comment]
Adding to the "this is a problem" discourse. These metaframeworks all using folder based routing makes us need this.

I've hacked the CSS to make the folder path on a second line. Doesn't cover every use case, but seems like an easy win especially as vertical monitor space is often easier to come by than horizontal. 

![image](https://github.com/microsoft/vscode/assets/176013/351e88b4-54d3-4b3e-aaa7-2eda00d65427)

Also, the custom API from this PR looks fantastic: https://github.com/microsoft/vscode/pull/197205

**Comment by leejoramo-d51** [comment]
@wesbos I like your  two line approach. I don't think I have seen this suggestion before. 

At first  I was going to reply that on notebooks computers I feel more vertically constrained. But in thinking about all of the UI issues related to these tabs,  I see you are correct. We are in fact Horizontally constrained regardless screen proportions. Even if you have a super wide screen, and the tabs plenty of space, I think stacking them like this results in a better visual grouping. 

And on my desktop systems, I personally code on screens in Portrait orientation, so this would be a big win.

**Comment by sammosna** [comment]
Any news? 🙏🏻

**Comment by geekley** [comment]
> I can't believe views like the Settings Editor (UI) and Shortcut Editor still use a generic "file" icon instead of something like a cogwheel/wrench/etc.

It seems this is no longer the case on latest update :tada:
Does this also mean there's now an API for setting custom icons on tabs of custom editors too? :pray:

**Comment by benibenj** [comment]
In tomorrows insiders we will support custom editor labels which can be provided with the `workbench.editor.customLabels.patterns` setting. This setting allows users to to provide a (glob) **pattern** to match with file paths and a **template** which defines how the label should be rendered if the pattern matches.

https://github.com/microsoft/vscode/assets/44439583/cb159b20-5fa3-4d1e-b728-453e81f95f57

**Comment by karlhorky** [comment]
@benibenj thanks so much for this! 🎉

Will patterns with negative `dirname` indices be supported? (counting from end, similar to [the `Array.prototype.at()` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at)) Eg. `${dirname(-1)}` for the last directory:

```
root/fileA.txt                          # dirname(-1) -> root
root/folder1/fileB.txt                  # dirname(-1) -> folder1
root/folder1/folder2/folder3/fileC.txt  # dirname(-1) -> folder3

fileD.txt                              # dirname(-1) -> '' (empty string)
```

Examples with other negative numbers:

```
root/folder1/folder2/folder3/fileC.txt  # dirname(-3) -> folder1

fileD.txt                              # dirname(-2) -> '' (empty string)
```

**Comment by benibenj** [comment]
@karlhorky Negative `dirname()` indices are currently not supported. Would this be helpful?
Here an example to clarify how the dirname indices work:
```
Path: root/grandParentFolder/parentFolder/file.txt
- dirname(0) = dirname = parentFolder
- dirname(1) = grandParentFolder
- dirname(2) = root
```

**Comment by lramos15** [comment]
@karlhorky Look at the example @benibenj posted. The indexes already work in the "negative" direction you're thinking of. So `${dirname}` would be `ContactForm`

**Comment by lramos15** [comment]
@Luke-Callaghan23 We want to see first if these settings help before adding an API as an API is a much more complex undertaking. It appeared based on the issue history that a lot of cases will be solved by this setting. It does not mean that the PR will not happen, just that it will not come until later depending on user need.

**Comment by karlhorky** [comment]
> Look at the example @benibenj posted. The indexes already work in the "negative" direction you're thinking of

@lramos15 Ah great, thanks for the clarification and sorry for the noise! I missed this detail in the explanation from the screenshot (I thought it was 1-indexed) from the example.

I'll edit my posts to minimize the confusion.

**Comment by benibenj** [comment]
@wesbos Glob patterns do not support capture groups so this is currently not possible, however, it should be pretty easy to just add multiple patterns to cover the different cases. You can accomplish your dynamic routes problem with two different entries.

Regarding API: https://github.com/microsoft/vscode/issues/41909#issuecomment-2010283035

Something we are considering is for extensions to provide common patterns for the `workbench.editor.customLabels.patterns` setting. This would not require any API and would be much easier for extensions to support.

**Comment by wesbos** [comment]
Great thank you - both of those answers seem more well thought out than my suggestion - thanks a ton!

**Comment by benmccann** [comment]
> Something we are considering is for extensions to provide common patterns for the workbench.editor.customLabels.patterns setting. This would not require any API and would be much easier for extensions to support.

Can you clarify what this means? E.g. would VS Code ship with default patterns for SvelteKit, Next.js, etc.? When you say it would be easier for extensions to support, how would extensions be involved?

**Comment by mkalinski93** [comment]
Is it possible to somehow have parts of the names be upper/lower cased? e.g: I have an Index.vue file in a folder called "products". Then the custom label would be products.vue. I would like to have it Products.vue instead.

**Comment by chromey** [comment]
I want to use this in an nx monorepo so that every filename is prefixed by the name of the app, e.g.
`/apps/myapp/sub/dir/somefile.ts` should be displayed as `myapp: somefile.ts`. Am I right to assume that this is not possible given that the indexes are counted from the end of the path and files exist at any nesting level below the app?

**Comment by benibenj** [comment]
@chromey the current implementation does not support this. What you can do for the meantime is to an entry for each app. I can see it would make sense to better support such cases. What we could do is add support for negative indices which would allow accessing the folder names from the start of the path. Would this be helpful for you?

**Comment by chromey** [comment]
Thanks @benibenj, this would be helpful indeed. But the workaround you suggest is also an option for my case, given that there are only ~15 apps in the repo. Thanks again.

**Comment by benibenj** [comment]
@chromey I'll be adding support for this as I can see it is clearly needed. I'm tracking the issue here #209619. Will however be part of next months release

**Comment by benmccann** [comment]
Would it also be possible to support new lines? I'd love to be able to do something like https://github.com/microsoft/vscode/issues/41909#issuecomment-1875768035

**Comment by benibenj** [comment]
@benmccann Please create a feature request for this

**Comment by benmccann** [comment]
> @benmccann Please create a feature request for this

Thank you. Done here: https://github.com/microsoft/vscode/issues/209701

**Comment by Luke-Callaghan23** [comment]
Hey thanks for the effort on this issue.  It's great to see it getting added in this latest release.  I'm wondering now if there will ever be any effort/ability to work with the _hint_ label of a tab (as in the grayed out and italicized folder name next to the label)?  As far as I can tell, the only options to modify the _hint_ of tab label is `workbench.editor.labelFormat`.  For me, I'd like to be able hide the hint completely, but it still shows when: two tabs are open with the same label but have different direct parent directories.  

Ideally, we could have complete control like we have with the main label through something like `workbench.editor.customLabels.hintPatterns` but even just having a new option in `workbench.editor.labelFormat` for something like "Never show hints" would be fine for my situation.  Here is an example of what I want to avoid:

![image](https://github.com/microsoft/vscode/assets/56779108/fa556bec-68f2-40eb-9fa7-3f67875e09e8)

Duplicate labels inside of two different folders with very long names taking up the entire space of my tab bar.

**Comment by benmccann** [comment]
@benibenj do you think that https://github.com/microsoft/vscode/issues/209701 might be something the VS Code team is likely to address in the near future? If not, it's possible I may be able to have the Svelte team sponsor it

**Comment by riungemaina** [comment]
Hi. 
Any idea what I'm missing here? 
this is what I'm getting
```
Path: root/parentFolder/folder/file.txt
- ${dirname(0)} = ${dirname} = parentFolder
- ${dirname(1)} = root
```
![Screenshot 2024-04-26 at 19 22 58](https://github.com/microsoft/vscode/assets/51113345/b2f0b708-b204-4c28-b3c8-7d2d0d0048a4)
![Screenshot 2024-04-26 at 19 23 12](https://github.com/microsoft/vscode/assets/51113345/b5bdb516-fbc6-46c9-b22a-4a255ff308d4)

**Comment by benibenj** [comment]
@riungemaina that's a bug which was introduced in yesterdays insiders. Tracking in #211522

**Comment by Luke-Callaghan23** [comment]
Is this no longer working in the latest VSCode insiders?  None of the old patterns are drawing for me and even using full paths won't work.


https://github.com/microsoft/vscode/assets/56779108/f82d3090-98eb-4b7e-ba87-005ba03f7379

**Comment by ssalka** [comment]
Would be great if we were able to set custom titles not only for the file names but also the labels/hints (`workbench.editor.labelFormat`) – I work in a monorepo and would like to have the hint only show the sub-package the file resides in. I can sorta do that with custom format`[${dirname(-3)}] ${filename}`, but it doesn't look quite as nice since the package name doesn't have the same styling as the hint (smaller + greyed out text)

**Comment by benibenj** [comment]
@Luke-Callaghan23 Relative patterns now have to include the folder name itself. This means you should change your patterns to one of the following:
- `*/.vscode/settings.json`
- `_FOLDER_NAME_/.vscode/settings.json`

**Comment by babakfp** [comment]
> Awesome job on this! Just tried it out on a sveltekit and Next.js project. Some questions / possible feedback:
> 
> Is there any chance to access the capture groups in the result? Do globs even have capture groups? I found a few instances that maybe it could be helpful, but I could also think maybe we just write multiple globs for these instances.
> 
> `app/listings/page.tsx` is a common route, and `"**/app/**/page.tsx": "${dirname}.${extname}",` will give me `listings.tsx` - perfect.
> 
> But now I want to write one for dynamic routes:
> 
> `app/listings/[id]/page.tsx` and `app/albums/[id]/page.tsx`
> 
> So in this case `[id].tsx` isn't a great name for the tab. In this page I want `listings/[id].tsx` and `albums/[id].tsx`
> 
> Maybe another question for this is if this tab-naming API will be made available to extensions, so the Next.js/Sveltekit Plugins can dynamically/conditionally generate the tab name (and maybe icon?)

https://github.com/microsoft/vscode/issues/223107#issuecomment-2292519067

**Comment by firelizzard18** [comment]
@benibenj @lramos15 Another use case: providing a meaningful label for custom editors. In my extension I am generating CPU/memory usage profiles and storing them in the extension's storage directory (or the OS temp directory, depending) and using a custom editor to present those to the user. As such, the default label is meaningless to the user. If I were able to set the title of the custom editor, I could make that something meaningful to the user. Unfortunately setting the `title` property of the webview apparently has no effect. I'd prefer an API for this, maybe part of the custom editor provider, so I can provide a useful label without having to hack it into the filename.
